python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py

```diff
@@ -0,0 +1,184 @@
+            and self.asset_graph == other.asset_graph
+            and self.partitions_subsets_by_asset_key == other.partitions_subsets_by_asset_key
+        )
+
+    def __repr__(self) -> str:
+        return f"AssetGraphSubset(partitions_subset_by_asset_key={self.partitions_subsets_by_asset_key})"
+
+    def __len__(self) -> int:
+        return len(self._non_partitioned_asset_keys) + sum(
+            len(subset) for subset in self._partitions_subsets_by_asset_key.values()
+        )
+
+    @classmethod
+    def from_asset_partition_set(
+        cls, asset_partitions_set: AbstractSet[AssetKeyPartitionKey], asset_graph: AssetGraph
+    ) -> "AssetGraphSubset":
+        partitions_by_asset_key = defaultdict(set)
+        non_partitioned_asset_keys = set()
+        for asset_key, partition_key in asset_partitions_set:
+            if partition_key is not None:
+                partitions_by_asset_key[asset_key].add(partition_key)
+            else:
+                non_partitioned_asset_keys.add(asset_key)
+
+        return AssetGraphSubset(
+            partitions_subsets_by_asset_key={
+                asset_key: cast(PartitionsDefinition, asset_graph.get_partitions_def(asset_key))
+                .empty_subset()
+                .with_partition_keys(partition_keys)
+                for asset_key, partition_keys in partitions_by_asset_key.items()
+            },
+            non_partitioned_asset_keys=non_partitioned_asset_keys,
+            asset_graph=asset_graph,
+        )
+
+    @classmethod
+    def from_storage_dict(
+        cls, serialized_dict: Mapping[str, Any], asset_graph: AssetGraph
+    ) -> "AssetGraphSubset":
+        partitions_subsets_by_asset_key: Dict[AssetKey, PartitionsSubset] = {}
+        for key, value in serialized_dict["partitions_subsets_by_asset_key"].items():
+            asset_key = AssetKey.from_user_string(key)
+            partitions_def = cast(PartitionsDefinition, asset_graph.get_partitions_def(asset_key))
+            partitions_subsets_by_asset_key[asset_key] = partitions_def.deserialize_subset(value)
+
+        non_partitioned_asset_keys = {
+            AssetKey.from_user_string(key) for key in serialized_dict["non_partitioned_asset_keys"]
+        }
+
+        return AssetGraphSubset(
+            asset_graph, partitions_subsets_by_asset_key, non_partitioned_asset_keys
+        )
+
+    @classmethod
+    def all(cls, asset_graph: AssetGraph) -> "AssetGraphSubset":
+        partitions_subsets_by_asset_key: Dict[AssetKey, PartitionsSubset] = {}
+        non_partitioned_asset_keys: Set[AssetKey] = set()
+
+        for asset_key in asset_graph.all_asset_keys:
+            partitions_def = asset_graph.get_partitions_def(asset_key)
+            if partitions_def:
+                partitions_subsets_by_asset_key[
+                    asset_key
+                ] = partitions_def.empty_subset().with_partition_keys(
+                    partitions_def.get_partition_keys()
```

Comment:
There's definitely room to optimize this to be O(1) instead of O(N partitions). My thought was to leave that for future work because right now this is just used in tests, with small # partitions.